crda/remotematcher.go

```diff
@@ -0,0 +1,404 @@
+	packageVersionToIndexRecord := make(map[string]*claircore.IndexRecord)
+	key := func(Name, Version string) string {
+		return fmt.Sprintf("%s@%s", Name, Version)
+	}
+	for _, ir := range records {
+		packageVersionToIndexRecord[key(ir.Package.Name, ir.Package.Version)] = ir
+	}
+
+	results := make(map[string][]*claircore.Vulnerability)
+	ctrlC, errorC := m.invokeComponentAnalysesInBatch(ctx, records)
+	err := <-errorC // guaranteed to have an err or be closed.
+	// Don't propagate error, log and move on.
+	if err != nil {
+		zlog.Error(ctx).Err(err).Msg("remote api call failure")
+		return results, nil
+	}
+	for vrs := range ctrlC {
+		for _, vr := range vrs {
+			ir := packageVersionToIndexRecord[key(vr.Name, vr.Version)]
+
+			// A package can have 0 or more vulnerabilities for a version.
+			var vulns []*claircore.Vulnerability
+			for _, vuln := range vr.Vulnerabilities {
+				vulns = append(vulns, &claircore.Vulnerability{
+					ID:                 vuln.ID,
+					Updater:            "CodeReadyAnalytics",
+					Name:               vuln.ID,
+					Description:        vuln.Title,
+					Links:              vuln.URL,
+					Severity:           vuln.Severity,
+					NormalizedSeverity: NormalizeSeverity(vuln.Severity),
+					FixedInVersion:     strings.Join(vuln.FixedIn, ", "),
+					Package:            ir.Package,
+					Repo:               ir.Repository,
+				})
+			}
+			results[ir.Package.ID] = append(results[ir.Package.ID], vulns...)
+		}
+	}
+	zlog.Debug(ctx).
+		Int("vulnerabilities", len(results)).
+		Msg("response")
+	return results, nil
+}
+
+func (m *Matcher) invokeComponentAnalysesInBatch(ctx context.Context, records []*claircore.IndexRecord) (<-chan []*VulnReport, <-chan error) {
+	inC := make(chan []*claircore.IndexRecord, m.requestConcurrency)
+	ctrlC := make(chan []*VulnReport, m.requestConcurrency)
+	errorC := make(chan error, 1)
+	batchSize := m.batchSize
+	go func() {
+		defer close(errorC)
+		defer close(ctrlC)
+		var g errgroup.Group
+		for start := 0; start < len(records); start += batchSize {
+			end := start + batchSize
+			if end > len(records) {
+				end = len(records)
+			}
+			g.Go(func() error {
+				vulns, err := m.invokeComponentAnalyses(ctx, <-inC)
+				if err != nil {
+					return err
+				}
+				ctrlC <- vulns
+				return nil
+			})
+			inC <- records[start:end]
+		}
+		close(inC)
+		if err := g.Wait(); err != nil {
+			errorC <- err
+		}
+	}()
+	return ctrlC, errorC
+}
+
+func (m *Matcher) invokeComponentAnalyses(ctx context.Context, records []*claircore.IndexRecord) ([]*VulnReport, error) {
+	// prepare request.
+	request := VulnRequest{
+		Ecosystem: m.ecosystem,
+		Packages:  make([]Package, len(records)),
+	}
+	for i, ir := range records {
+		request.Packages[i] = Package{
+			Name:    ir.Package.Name,
+			Version: ir.Package.Version,
+		}
+	}
+	// A request shouldn't go beyound 5s.
+	tctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+	defer cancel()
+	reqBody, _ := json.Marshal(request)
```

Comment:
Check the error.

Reply:
ACK.

Reply:
@rafiu007 like this I believe
```go
	reqBody, err := json.Marshal(request)
        if err != nil {
             // handle error
        }
	req, err := http.NewRequestWithContext(tctx, http.MethodPost, m.url.String(), bytes.NewBuffer(reqBody))
        if err != nil {
            // handle error
        }
```
(substitute the json.Marshal for the json.Decoder implementation)

Reply:
Got it. My bad, I misunderstood it for a different part of the code. :)

Reply:
Done.
Resolving this conversation.